File: app/src/main/java/com/rolandvitezhu/todocloud/network/ApiService.java
package com.rolandvitezhu.todocloud.network;

import com.rolandvitezhu.todocloud.network.api.category.dto.GetCategoriesResponse;
import com.rolandvitezhu.todocloud.network.api.category.dto.InsertCategoryRequest;
import com.rolandvitezhu.todocloud.network.api.category.dto.InsertCategoryResponse;
import com.rolandvitezhu.todocloud.network.api.category.dto.UpdateCategoryRequest;
import com.rolandvitezhu.todocloud.network.api.category.dto.UpdateCategoryResponse;
import com.rolandvitezhu.todocloud.network.api.list.dto.GetListsResponse;
import com.rolandvitezhu.todocloud.network.api.list.dto.InsertListRequest;
import com.rolandvitezhu.todocloud.network.api.list.dto.InsertListResponse;
import com.rolandvitezhu.todocloud.network.api.list.dto.UpdateListRequest;
import com.rolandvitezhu.todocloud.network.api.list.dto.UpdateListResponse;
import com.rolandvitezhu.todocloud.network.api.todo.dto.GetTodosResponse;
import com.rolandvitezhu.todocloud.network.api.todo.dto.InsertTodoRequest;
import com.rolandvitezhu.todocloud.network.api.todo.dto.InsertTodoResponse;
import com.rolandvitezhu.todocloud.network.api.todo.dto.UpdateTodoRequest;
import com.rolandvitezhu.todocloud.network.api.todo.dto.UpdateTodoResponse;
import com.rolandvitezhu.todocloud.network.api.user.dto.LoginUserRequest;
import com.rolandvitezhu.todocloud.network.api.user.dto.LoginUserResponse;
import com.rolandvitezhu.todocloud.network.api.user.dto.ModifyPasswordRequest;
import com.rolandvitezhu.todocloud.network.api.user.dto.ModifyPasswordResponse;
import com.rolandvitezhu.todocloud.network.api.user.dto.RegisterUserRequest;
import com.rolandvitezhu.todocloud.network.api.user.dto.RegisterUserResponse;
import com.rolandvitezhu.todocloud.network.api.user.dto.ResetPasswordRequest;
import com.rolandvitezhu.todocloud.network.api.user.dto.ResetPasswordResponse;

import io.reactivex.Single;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;

public interface ApiService {

  @POST("todo_cloud/v1/user/register")
  Single<RegisterUserResponse> registerUser(@Body RegisterUserRequest request);

  @POST("todo_cloud/v1/user/login")
  Single<LoginUserResponse> loginUser(@Body LoginUserRequest request);

  @POST("todo_cloud/v1/user/reset_password")
  Single<ResetPasswordResponse> resetPassword(@Body ResetPasswordRequest request);

  @POST("todo_cloud/v1/user/modify_password")
  Single<ModifyPasswordResponse> modifyPassword(@Body ModifyPasswordRequest request);

  @GET("todo_cloud/v1/todo/{row_version}")
  Single<GetTodosResponse> getTodos(@Path("row_version") int rowVersion);

  @POST("todo_cloud/v1/todo/insert")
  Single<InsertTodoResponse> insertTodo(@Body InsertTodoRequest request);

  // 000webhost.com doesn't allow PUT and DELETE requests for free accounts
  @POST("todo_cloud/v1/todo/update")
  Single<UpdateTodoResponse> updateTodo(@Body UpdateTodoRequest request);

  @GET("todo_cloud/v1/list/{row_version}")
  Single<GetListsResponse> getLists(@Path("row_version") int rowVersion);

  @POST("todo_cloud/v1/list/insert")
  Single<InsertListResponse> insertList(@Body InsertListRequest request);

  // 000webhost.com doesn't allow PUT and DELETE requests for free accounts
  @POST("todo_cloud/v1/list/update")
  Single<UpdateListResponse> updateList(@Body UpdateListRequest request);

  @GET("todo_cloud/v1/category/{row_version}")
  Single<GetCategoriesResponse> getCategories(@Path("row_version") int rowVersion);

  @POST("todo_cloud/v1/category/insert")
  Single<InsertCategoryResponse> insertCategory(@Body InsertCategoryRequest request);

  // 000webhost.com doesn't allow PUT and DELETE requests for free accounts
  @POST("todo_cloud/v1/category/update")
  Single<UpdateCategoryResponse> updateCategory(@Body UpdateCategoryRequest request);
}

File: app/src/main/java/com/rolandvitezhu/todocloud/network/ApiService.kt
package com.rolandvitezhu.todocloud.network

import com.rolandvitezhu.todocloud.data.Category
import com.rolandvitezhu.todocloud.data.List
import com.rolandvitezhu.todocloud.data.Todo
import com.rolandvitezhu.todocloud.data.User
import com.rolandvitezhu.todocloud.network.api.category.dto.GetCategoriesResponse
import com.rolandvitezhu.todocloud.network.api.dto.BaseResponse
import com.rolandvitezhu.todocloud.network.api.list.dto.GetListsResponse
import com.rolandvitezhu.todocloud.network.api.rowversion.dto.GetNextRowVersionResponse
import com.rolandvitezhu.todocloud.network.api.todo.dto.GetTodosResponse
import com.rolandvitezhu.todocloud.network.api.user.dto.LoginUserResponse
import com.rolandvitezhu.todocloud.network.api.user.dto.ModifyPasswordRequest
import retrofit2.http.*

interface ApiService {
    @POST("todo_cloud/v3/user/register")
    suspend fun registerUser(@Body request: User): BaseResponse

    @POST("todo_cloud/v3/user/login")
    suspend fun loginUser(@Body request: User): LoginUserResponse

    @POST("todo_cloud/v3/user/reset_password")
    suspend fun resetPassword(@Body request: User): BaseResponse

    @POST("todo_cloud/v3/user/modify_password")
    suspend fun modifyPassword(@Body request: ModifyPasswordRequest): BaseResponse

    @GET("todo_cloud/v3/todo/{row_version}")
    suspend fun getTodos(@Path("row_version") rowVersion: Int): GetTodosResponse

    @POST("todo_cloud/v3/todo/insert")
    suspend fun insertTodo(@Body request: Todo): BaseResponse

    // 000webhost.com does not allow PUT and DELETE requests for free accounts
    @POST("todo_cloud/v3/todo/update")
    suspend fun updateTodo(@Body request: Todo): BaseResponse

    @GET("todo_cloud/v3/list/{row_version}")
    suspend fun getLists(@Path("row_version") rowVersion: Int): GetListsResponse

    @POST("todo_cloud/v3/list/insert")
    suspend fun insertList(@Body request: List): BaseResponse

    // 000webhost.com does not allow PUT and DELETE requests for free accounts
    @POST("todo_cloud/v3/list/update")
    suspend fun updateList(@Body request: List): BaseResponse

    @GET("todo_cloud/v3/category/{row_version}")
    suspend fun getCategories(@Path("row_version") rowVersion: Int): GetCategoriesResponse

    @POST("todo_cloud/v3/category/insert")
    suspend fun insertCategory(@Body request: Category): BaseResponse

    // 000webhost.com does not allow PUT and DELETE requests for free accounts
    @POST("todo_cloud/v3/category/update")
    suspend fun updateCategory(@Body request: Category): BaseResponse

    @GET("todo_cloud/v3/get_next_row_version/{table}")
    suspend fun getNextRowVersion(
            @Path("table") table: String?, @Header("authorization") apiKey: String?
    ): GetNextRowVersionResponse
}